### PR TITLE
Feature/v2: action-state-repo-generate-pr >> action-state-repo-update-image

### PR DESCRIPTION
**Pull request by AlbertoFemenias**
- Create pull request with automerge transformed into update image + cretate pr with automerge.

- If you wish to test this action you can use repo prefapp/test-action > actions > test

- The purpose of this action is to change the images values (state_repo/app/env/images.yaml.service.image) stored in the state repo for 1 or multiple services

- The entry point is index.js which uses the utils classes:
  - ghUtils
  - inputUtils
  - yamlUtils

- Each component has its own unit tests

- The node dependencies required are:
  - "@actions/core": "^1.2.5",
  - "@actions/exec": "^1.1.0",
  - "js-yaml": "^4.1.0",
  - "sha1": "^1.1.1"

- This action now runs as follows:
  1. Checks:
    - exists `/config.yml`?
      - exists auto-merge config for environment ?
         - if it does not exists: **auto-merge**: false
   - exists path: `<application>/<environmet>/images.yml`?
      - foreach <service>: exists key images.yaml.<service>
  
  2. Modify image (for each service)
  3. Commit changes
  4. Create PR with body, title & reviewers
  5. Automerge if possible (usually dev and pre)

- Usage example:
```yaml
steps:
  - uses: actions/checkout@v2  #NECESSARY
  - name: Update image  
      uses: prefapp/action-state-repo-update-image@feature/v2
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} #NECESSARY
      with:
        application: appfoo
        environment: dev/pre/pro
        services: svc123
        image: image:new
        reviewers: juana, suhermana #OPTIONAL
        pr_title: PR title #OPTIONAL (not recomended) 
        pr_body: PR body #OPTIONAL (not recomended) 
        branch_name: automated/update-image #OPTIONAL

```

### CODE REVIEW

File: utils/inputUtils.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  static createBranchName(app, env){
+    
+    const rndSha = sha1(Date.now()).slice(0, 6);
+    return `automated/update-image-${app}-${env}-${rndSha}`;
+    
+  }
+
+}
+
+module.exports = inputUtils;
```
Comment (alambike) [inline utils/inputUtils.js:23]: Missing ending empty newline.

File: utils/yamlUtils.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+const yaml = require('js-yaml');
+const fs   = require('fs');
+
+
```
Comment (alambike) [inline utils/yamlUtils.js:4]: remove extra blank line.

File: utils/yamlUtils.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    for (let i = 0; i < services.length; i++){
+      oldImages = yamlUtils.modifyImage(application, environment, services[i], newImage);
+    }
+
+    return oldImages;
+
+  }
+}
+
+module.exports = yamlUtils;
```
Comment (alambike) [inline utils/yamlUtils.js:67]: Empty ending line

**Review by alambike — APPROVED** [review]

**Review by frmadem — APPROVED** [review]

File: fixtures/app1/pro/images.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+app-server:
+  image: foo/server:pro
+
+app-client:
+  image: foo/client:pro
+
+app-api:
+  image: foo/api:pro
```
Comment (jotasixto) [inline fixtures/app1/pro/images.yaml]: No newline at end of file

File: fixtures/app2/dev/images.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+proxy:
+  image: foo/proxy:dev
+
+dns:
+  image: foo/dns:dev
```
Comment (jotasixto) [inline fixtures/app2/dev/images.yaml]: No newline at end of file

File: fixtures/app2/pre/images.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+proxy:
+  image: foo/proxy:pre
+
+dns:
+  image: foo/dns:pre
```
Comment (jotasixto) [inline fixtures/app2/pre/images.yaml]: No newline at end of file

File: fixtures/config.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    - dns
+
+app3:
+  auto_merge:
+    dev: false
+    pre: true
+    pro: false
+  services:
+    - bbdd
+    - soap
```
Comment (jotasixto) [inline fixtures/config.yaml]: No newline at end of file

**Review by jotasixto — APPROVED** [review]
LGTM!

**Pull request merged**